### PR TITLE
Moves request error handling to user api request

### PR DESCRIPTION
**Pull request by diogob**
The idea behind this refactor is to extract error cases that were inside the `ApiRequest` type (namely as the `ActionInappropriate` and `PayloadParseError` constructors).

Now this error cases are handled by the `userApiRequest` that returns an `Either ApiRequestError ApiRequest`.

The type `ApiRequestError` has a function inside the error module that converts this type into an HTTP response. This function is used in the `postgrest` function making the `app` function slightly simpler as well.

### CODE REVIEW

File: src/PostgREST/QueryBuilder.hs [hunk 1]
```diff
@@ -113,7 +113,6 @@ createReadStatement selectQuery countQuery isSingle countTotal asCsv =
 createWriteStatement :: QualifiedIdentifier -> SqlQuery -> SqlQuery -> Bool ->
                         PreferRepresentation -> [Text] -> Bool -> Payload ->
                         H.Query UniformObjects (Maybe ResultsWithCount)
-createWriteStatement _ _ _ _ _ _ _ (PayloadParseError _) = undefined
```
Comment (ruslantalpa) [inline src/PostgREST/QueryBuilder.hs]: i like how these are going away :)

File: src/PostgREST/ApiRequest.hs [hunk 2]
```diff
@@ -113,84 +117,79 @@ data ApiRequest = ApiRequest {
...
-     . parseSimpleQuery
-     $ rawQueryString req
-  , iJWT = tokenStr
-  }
-
+userApiRequest :: Schema -> Request -> RequestBody -> Either ApiRequestError ApiRequest
+userApiRequest schema req reqBody
+  | isTargetingProc && method /= "POST" = Left ErrorActionInappropriate
+  | topLevelRange == emptyRange = Left ErrorInvalidRange
+  | shouldParsePayload && isLeft payload = either (Left . ErrorInvalidBody . toS) undefined payload
```
Comment (begriffs) [inline src/PostgREST/ApiRequest.hs]: This is good! Lewis Stern just told us in gitter about a problem that turned out to be caused because postgrest didn't issue a descriptive error for a malformed post body. Your change will prevent other people from getting confused.

Could you add a test that posting a malformed body returns an error? Lewis' particular problem was sending a JSON object quoted as a string rather than as a real object. I'm not sure your code would catch that case because a string is valid json. Postgrest will probably treat it as an empty object, which is confusing for the user.

File: src/PostgREST/ApiRequest.hs [hunk 3]
```diff
@@ -113,84 +117,79 @@ data ApiRequest = ApiRequest {
...
+  topLevelRange = fromMaybe allRange $ M.lookup "limit" ranges
+  action =
+    if isTargetingProc
+      then ActionInvoke
+      else
+        case method of
+            "GET"     -> if target == TargetRoot
+                          then ActionInspect
+                          else ActionRead
+            "POST"    -> ActionCreate
```
Comment (begriffs) [inline src/PostgREST/ApiRequest.hs]: What about putting `if isTargetingProc` inside this POST case to avoid indenting the whole case statement?

File: src/PostgREST/ApiRequest.hs [hunk 4]
```diff
@@ -113,84 +117,79 @@ data ApiRequest = ApiRequest {
...
+  target = case path of
+              []            -> TargetRoot
+              [table]       -> TargetIdent
+                              $ QualifiedIdentifier schema table
+              ["rpc", proc] -> TargetProc
+                              $ QualifiedIdentifier schema proc
+              other         -> TargetUnknown other
+  shouldParsePayload = action `elem` [ActionCreate, ActionUpdate, ActionInvoke]
+  relevantPayload = if shouldParsePayload
+                      then case payload of
```
Comment (begriffs) [inline src/PostgREST/ApiRequest.hs]: Check out `rightToMaybe` in the either package.

File: src/PostgREST/App.hs [hunk 5]
```diff
@@ -81,17 +80,18 @@ postgrest conf refDbStructure pool getTime =
...
+        let eClaims = jwtClaims
+              (secret <$> configJwtSecret conf) (iJWT apiRequest) time
+            authed = containsRole eClaims
+            handleReq = runWithClaims conf eClaims (app dbStructure conf) apiRequest
+            txMode = transactionMode $ iAction apiRequest
 
-    resp <- either (pgErrResponse authed) id <$> P.use pool
-      (HT.run handleReq HT.ReadCommitted txMode)
-    respond resp
+        resp <- either (pgErrResponse authed) id <$> P.use pool
```
Comment (begriffs) [inline src/PostgREST/App.hs]: `mapLeft` from either package

File: src/PostgREST/QueryBuilder.hs (hunk 1 above)
Comment (begriffs) [inline src/PostgREST/QueryBuilder.hs]: +1

File: src/PostgREST/Types.hs [hunk 6]
```diff
@@ -111,9 +110,7 @@ unUniformObjects (UniformObjects objs) = objs
 -- | When Hasql supports the COPY command then we can
 -- have a special payload just for CSV, but until
 -- then CSV is converted to a JSON array.
-data Payload = PayloadJSON UniformObjects
-             | PayloadParseError BS.ByteString
-             deriving (Show, Eq)
+data Payload = PayloadJSON UniformObjects deriving (Show, Eq)
```
Comment (begriffs) [inline src/PostgREST/Types.hs]: Is there a reason to keep Payload as a type now that there is only one constructor? Perhaps we could use UniformObjects throughout the code instead.

Comment (diogob) [inline src/PostgREST/Types.hs]: True, I actually only refrained from doing this to avoid too many steps in one PR 😛 
I'll include this

File: src/PostgREST/ApiRequest.hs (hunk 2 above)
Comment (diogob) [inline src/PostgREST/ApiRequest.hs]: Sure, I'll check that out.

File: src/PostgREST/ApiRequest.hs (hunk 3 above)
Comment (diogob) [inline src/PostgREST/ApiRequest.hs]: I like moving the if inside the case, it also creates a nice symmetry with the GET case.
If we need a different mapping later one (1 Action -> Multiple verbs) we probably will want to rethink this section of the code anyway.

File: src/PostgREST/ApiRequest.hs (hunk 2 above)
Comment (diogob) [inline src/PostgREST/ApiRequest.hs]: @begriffs from what I've seen we would still accept an invalid PayloadJSON as long as is a valid JSON.
I'd love to add that constraint but I believe that would be more suited to another pull request, since this is becoming quite large already.

**Comment by begriffs** [comment]
Thank you!
